### PR TITLE
Properly deserialize context when deserializing state in the entity state cache

### LINKED ISSUE #1156: All values in Context are null or empty in StateChange objects
## The problem
It appears the lazy deserialization of StateChanges in 24.33.0 broke the deserialization of the Context.  In both the New and Old properties the Context has an empty string for the Id and null for ParentId and UserId.
image

## Environment
Running NetDaemon in Docker using the code in this repo [nuttytree-netdaemon](https://github.com/nuttytree/nuttytree-netdaemon) connected to Home Assistant version 2024.8.2 running in Docker.

- NetDaemon release with the issue: 24.33.0
- Last working NetDaemon release (if known): 24.28.1
- Operating environment (Home assistant Add-on/Docker/Dev setup): Docker
- Link to integration documentation on our website:

## Link to or paste code that causes the issue
Looks like the issue is the result of this pull request #1146 
```c#

```

## Traceback/Error logs
```txt

```

### PR DESCRIPTION
**Pull request by nuttytree**
## Proposed change
Add JsonPropertyName attributes to properties in the Context object so that they properly deserialize when EntityStateCache deserializes the State object.

## Type of change
- [ ] Dependency upgrade
- [x] Bugfix (non-breaking change which fixes an issue)
- [ ] New feature (which adds functionality to an existing integration)
- [ ] Breaking change (fix/feature causing existing functionality to break)
- [ ] Code quality improvements to existing code or addition of tests

## Additional information
- This PR fixes or closes issue: fixes #1156 

## Checklist
- [x] The code change is tested and works locally.
- [x] Local tests pass. **Your PR cannot be merged unless tests pass**
- [x] The code compiles without warnings (code quality check)
- [x] Tests have been added to verify that the new code works.

### CODE REVIEW

**Review by FrankBakkerNl — CHANGES_REQUESTED** [review]
I think the fix is good, was not aware of this issue. What was the actual behavior you found?

It must have been introduced with the recent lazy deserialisation of state changes.

I think the tests you updated do not really validate this fix. I even expect them to also pass without the fix because they do not depend not depend. A better test would be to deserialize actually JSON data with the names with _ in them.

**Comment by Barsonax** [comment]
> It must have been introduced with the recent lazy deserialisation of state changes.

If this worked before its probably because JsonSerializerOptions used with with JsonNamingPolicy.CamelCase. If that is the case then the proper fix is to pass in those options again.

File: src/HassModel/NetDaemon.HassModel.Tests/Internal/EntityStateCacheTest.cs [hunk 1]
```diff
@@ -71,6 +75,8 @@ public async Task StateChangeEventIsFirstStoredInCacheThanForwarded()
         // Assert
         eventObserverMock.Verify(m => m.OnNext(It.IsAny<HassEvent>()));
         cache.GetState(entityId)!.State.Should().Be("newState");
+        cache.GetState(entityId)!.Context!.Id.Should().NotBeNullOrEmpty();
```
Comment (helto4real) [inline src/HassModel/NetDaemon.HassModel.Tests/Internal/EntityStateCacheTest.cs:78]: Why not just test the actual value?

**Comment by helto4real** [comment]
Weird that this have never worked, seeing your fix it is obvious.

File: src/HassModel/NetDaemon.HassModel.Tests/Internal/EntityStateCacheTest.cs (hunk 1 above)
Comment (nuttytree) [reply src/HassModel/NetDaemon.HassModel.Tests/Internal/EntityStateCacheTest.cs]: Updated to test the actual value is the expected value.

**Comment by nuttytree** [comment]
I have updated the tests with the requested changes.

**Review by FrankBakkerNl — APPROVED** [review]

File: src/HassModel/NetDaemon.HassModel.Tests/Internal/EntityStateCacheTest.cs (hunk 1 above)
[the remark this answers is not in the document]
Comment (nuttytree) [reply src/HassModel/NetDaemon.HassModel.Tests/Internal/EntityStateCacheTest.cs]: I meant to change all of those but missed that one, I updated that assert also.

**Comment by helto4real** [comment]
@nuttytree thank you for your contribution. Please make sure you sign the commits in the future. It is for everyones benefit and security.

**Pull request merged**